onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -15,23 +15,48 @@ namespace onnxruntime {
    etc.
    */
 
-// ComputeOffset: compute offset into a tensor. This is essentially the dot-product of
-// index and stride, restricted to the specified number of axes.
-static inline size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t>& stride, int64_t num_axes) {
-  size_t offset = 0;
-  for (int64_t j = 0; j < num_axes; ++j) {
-    offset += index[j] * stride[j];
+typedef struct MultiIndex {
+  size_t index;
+  size_t upper_bound;
+  int64_t stride;
+  MultiIndex() {}
+  void init(size_t i, size_t n, int64_t s) {
+    index = i;
+    upper_bound = n;
+    stride = s;
   }
-  return offset;
+} MultiIndex;
+
+static size_t IncrementIndexAndComputeOffsetSetup(MultiIndex* mindex, int64_t num_axes, const std::vector<int64_t>& target_dims,
+                                                  const std::vector<size_t>& stride, size_t element_size) {
+  size_t naxes = 0;
+  for (int64_t i = 0; i < num_axes; ++i) {
+    if (target_dims[i] == 1)
+      continue;
+    mindex[naxes].init(0, static_cast<size_t>(target_dims[i]), stride[i] * element_size);
+    ++naxes;
+  }
+  ORT_ENFORCE(naxes > 0, "Method IncrementIndexAndComputeOffset assumes this value is strictly positive.");
+  return naxes;
```

Comment:
This check should be in IncrementIndexAndComputeOffset if needed so it's done by the code that has the requirement for the value to be greater than zero.